### PR TITLE
Validate login form whever the input is changed

### PR DESCRIPTION
**Pull request by dukenv0307**
### Details

Validate login form whever the input is changed
### Fixed Issues

$ https://github.com/Expensify/App/issues/29640
PROPOSAL: https://github.com/Expensify/App/issues/29640#issuecomment-1763734530

### Tests

1. Log out the account
2. On the login page, enter the invalid email and click on the next button
3. Verify that the error appears
4. Change a letter from lowercase to uppercase or vice versa
5. Verify that the error still appears
- [x] Verify that no errors appear in the JS console

### QA Steps

1. Log out the account
2. On the login page, enter the invalid email and click on the next button
3. Verify that the error appears
4. Change a letter from lowercase to uppercase or vice versa
5. Verify that the error still appears
- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/Expensify/App/assets/129500732/49478460-9ed7-4680-8899-a48a4f6775d0

</details>

<details>
<summary>Android: mWeb Chrome</summary>

https://github.com/Expensify/App/assets/129500732/ea092750-b4f1-4448-81ca-b262de075574

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/Expensify/App/assets/129500732/905cb621-be38-41be-abad-2e393bdab166

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/Expensify/App/assets/129500732/c309a0e6-09d3-4157-ab6c-dd44d40ca20c

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/Expensify/App/assets/129500732/41bcad9b-e3a0-4eff-8cf5-e201151e46b4

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/Expensify/App/assets/129500732/4f25b04b-6726-44ad-a060-7060a0b99734

</details>

### CODE REVIEW

**Comment by Beamanator** [comment]
> > @aimane-chnaif It user `onMouseDown` event to prevent this issue. But the google icon is rendered by window.google how can we config that?
> 
> yeah, I am also concerned about that. I think we should raise discussion on the expected behavior.
> 
> cc: @Beamanator

I like the idea of bringing this to slack, though we can probably get away with validating on mouse down of the container that holds the sign in buttons, no? I wonder if the event bubbles up from the google icon to the container?

File: src/components/SignInButtons/GoogleSignIn/index.website.js [hunk 1]
```diff
@@ -78,7 +78,7 @@ function GoogleSignIn({translate, isDesktopFlow}) {
                 id={desktopId}
                 role={CONST.ACCESSIBILITY_ROLE.BUTTON}
                 aria-label={translate('common.signInWithGoogle')}
-                onMouseDown={(event) => event.preventDefault()}
+                // onMouseDown={(event) => event.preventDefault()}
```
Comment (Beamanator) [inline src/components/SignInButtons/GoogleSignIn/index.website.js]: I think we should remove this if we're not going to use it, yeah?

File: src/components/SignInButtons/GoogleSignIn/index.website.js [hunk 2]
```diff
@@ -88,7 +88,7 @@ function GoogleSignIn({translate, isDesktopFlow}) {
                 id={mainId}
                 role={CONST.ACCESSIBILITY_ROLE.BUTTON}
                 aria-label={translate('common.signInWithGoogle')}
-                onMouseDown={(event) => event.preventDefault()}
+                // onMouseDown={(event) => event.preventDefault()}
```
Comment (Beamanator) [inline src/components/SignInButtons/GoogleSignIn/index.website.js]: same

File: src/pages/signin/LoginForm/BaseLoginForm.js [hunk 3]
```diff
@@ -301,22 +302,26 @@ function LoginForm(props) {
...
-                                        style={[styles.textLabelSupporting, styles.textAlignCenter, styles.mb3, styles.mt2]}
-                                    >
-                                        {props.translate('common.signInWith')}
-                                    </Text>
-
-                                    <View style={props.isSmallScreenWidth ? styles.loginButtonRowSmallScreen : styles.loginButtonRow}>
-                                        <AppleSignIn />
-                                        <GoogleSignIn />
+                            CONFIG.ENVIRONMENT !== CONST.ENVIRONMENT.DEV ||
+                                (true && (
```
Comment (Beamanator) [inline src/pages/signin/LoginForm/BaseLoginForm.js:306]: What's the point of `true &&` here?

**Review by Beamanator — CHANGES_REQUESTED** [review]
A few little changes before we move forward

File: src/components/SignInButtons/GoogleSignIn/index.website.js (hunk 1 above)
Comment (Beamanator) [reply src/components/SignInButtons/GoogleSignIn/index.website.js]: Oh i might have commented on a specific commit, oops!!

File: src/pages/signin/LoginForm/BaseLoginForm.js [hunk 4]
```diff
@@ -140,35 +174,30 @@ function LoginForm(props) {
...
         }
 
-        const loginTrim = login.trim();
-        if (!loginTrim) {
-            setFormError('common.pleaseEnterEmailOrPhoneNumber');
-            return;
+        // For native, the single input doesn't lost focus when we click outside.
+        // So we need to change firstBlurred here to make the validate function is called whenever the text input is changed after the first validation.
+        if (!firstBlurred.current) {
+            firstBlurred.current = true;
```
[the remark this answers is not in the document]
Comment (Beamanator) [reply src/pages/signin/LoginForm/BaseLoginForm.js:181]: @aimane-chnaif it looks like this thread is still open, was this resolved?

**Comment by Beamanator** [comment]
Code is looking pretty good to me, though I'm not sure why the PR Reviewer Checklist is failing.. Seems like a bug since it looks like it got updated recently.

@aimane-chnaif will you please do a final retest before we move forwared?

**Review by aimane-chnaif — APPROVED** [review]
LGTM 🎉 
@Beamanator all yours

File: src/pages/signin/LoginForm/BaseLoginForm.js [hunk 5]
```diff
@@ -140,35 +175,30 @@ function LoginForm(props) {
             CloseAccount.setDefaultData();
         }
 
-        const loginTrim = login.trim();
-        if (!loginTrim) {
-            setFormError('common.pleaseEnterEmailOrPhoneNumber');
-            return;
+        // For native, the single input doesn't lost focus when we click outside.
+        // So we need to change firstBlurred here to make the validate function is called whenever the text input is changed after the first validation.
```
Comment (Beamanator) [inline src/pages/signin/LoginForm/BaseLoginForm.js:179]:
```suggestion
        // So we need to change firstBlurred here to make sure the validate function is called whenever the text input is changed after the first validation.
```

**Review by Beamanator — APPROVED** [review]
LGTM (one tiny comment copy change can be done later)

**Comment by roryabraham** [comment]
This is working locally, but not working as expected on staging: https://github.com/Expensify/App/issues/31413

**Pull request merged**